exploration/bidi-usability.md

```diff
@@ -348,47 +407,51 @@ the results or debug what is wrong with their messages.
-   allow for its proper isolation.
+```abnf
+; strongly directional marks and bidi isolates
+; ALM / LRM / RLM / LRI / RLI / FSI / PDI
+bidi = %x061C / %x200E / %x200F / %x2066-2069
 
-Quoted patterns, quoted literals, and names may be isolated by LRI/RLI/FSI...PDI.
-For names and quoted literals, the isolate characters are outside the body of the token,
-but for quoted patterns, the isolates are in the middle of the `{{` and `}}` characters.
-This avoids adding a lookahead requirement for detecting a `complex-message` start,
-and differentiates a `quoted-pattern` from a `quoted` `key` in a `variant`.
+; optional whitespace
+owsp = *( s / bidi )
 
-Expressions and markup may be isolated by LRI...PDI immediately within the `{` and `}`.
+; required whitespace
+wsp = [ owsp ] 1*s [ owsp ]
 
-An LRI is allowed immediately after a newline outside patterns and within expressions.
-This is intended to allow left-to-right representation for "code"
-even if it contains a newline followed by content
-that could otherwise prompt the paragraph direction to be detected as right-to-left.
+; whitespace characters
+s = ( SP / HTAB / CR / LF / %x3000 )
+```
 
-```abnf
-name           = [open-isolate] name-start *name-char [close-isolate]
-quoted         = [open-isolate] "|" *(quoted-char / quoted-escape) "|" [close-isolate]
-quoted-pattern = "{" [open-isolate] "{" pattern "}" [close-isolate] "}"
+**Pros**
+- Avoids problems with syntax errors that users and tools might find difficult to debug.
+- Effective if used carefully.
+- Addresses need to comply with UAX#31
 
-literal-expression    = "{" [LRI] [s] literal [s annotation] *(s attribute) [s] [close-isolate] "}"
-variable-expression   = "{" [LRI] [s] variable [s annotation] *(s attribute) [s] [close-isolate] "}"
-annotation-expression = "{" [LRI] [s] annotation *(s attribute) [s] [close-isolate] "}"
+**Cons**
+- Syntax does not prevent poor display outcomes, including enabling some Trojan Source cases (UAX#55);
+  note that tooling or linting can help ameliorate these issues.
 
-markup = "{" [LRI] [s] "#" identifier *(s option) *(s attribute) [s] ["/"] [close-isolate] "}"
-       / "{" [LRI] [s] "/" identifier *(s option) *(s attribute) [s] [close-isolate] "}"
+### Strict isolation all the time
+
+Apply bidi isolates in a strict way.
+The main differences to the proposed solution is:
+1. The open/close isolate characters are syntactically required to be paired.
+   This introduces parse errors for unpaired invisible characters,
+   which could lead to bad user experiences.
+
+As noted above, the "strict" version of the ABNF should be adopted by serializers and for 
+message normalization.
+
+// TODO put ABNF here
 
-s = 1*( SP / HTAB / CR / LF [LRI] / %x3000 )
-LRI = %x2066
-open-isolate  = %x2066-2068
-close-isolate = %x2069
-```
 
 Isolating rather than marking `name` helps ensure
 that its directionality does not spill over to adjoining syntax.
```

Comment:
What follows was a part of the "Loose Isolation" alternative, parts of which are in this PR included in the proposed design; this is separate from the "Strict isolation all the time" alternative, and should have its own title as well as the ABNF that was previously included.

````suggestion

### Isolate `name` rather than `unquoted-literal`

Isolating rather than marking `name` helps ensure
that its directionality does not spill over to adjoining syntax.

The following replaces the proposed design's changes to `literal` and the `[bidi]` additions to
`variable-expression`, `function`, `option`, `attribute`, `markup`, and `ns-separator`:
```abnf
name = [open-isolate] name-start *name-char [close-isolate]
quoted-literal = [open-isolate] "|" *(quoted-char / quoted-escape) "|" [close-isolate]
```

````